### ISSUE TITLE
Clarification needed on how parallelism works.

### ISSUE THREAD

**Issue by Sofocletus** [tdrussell/diffusion-pipe#330]
I'm currently testing GPU parallelism and I don't understand how it works, from the Readme:

"This code uses hybrid data- and pipeline-parallelism. Set the --num_gpus flag appropriately for your setup. Set pipeline_stages in the config file to control the degree of pipeline parallelism. Then the data parallelism degree will automatically be set to use all GPUs (number of GPUs must be divisible by pipeline_stages). For example, with 4 GPUs and pipeline_stages=2, you will run two instances of the model, each divided across two GPUs."

I'm currently running 4 GPUs with pipeline_stages=2, however the training speed seems similar to what I was getting with one GPU alone, although I can use bigger frame buckets because *I think* each pair of GPUs shares the VRAM. 

However I don't understand how the two paired (I suppose?) GPUs work together, what exactly is the meaning of "you will run two instances of the model, each divided across two GPU"?
Would that be similar to setting a batch of 2 when training? So that, for instance, a 50 epoch training run would have a similar effect as a 100 epoch one?
What I thought 4 GPUs with pipeline_stages=2 would do was to 1) allow me to use larger frame buckets and 2) double the speed of training.
Number 1 works OK, but I'm not seeing Number 2, so now I'm confused.

It's not clear to me, and considering it looks like the training I'm doing now will take days on Runpod, I'd really wouldn't want to be paying double if there's no benefit for it.

**Comment by Sofocletus** [comment]
If I may add my own question, just so anyone with the same doubts comes along.
Parallelism works, as far as I can see, as increasing the train batch, basically setting pipeline_stages=x works as setting train batch to x.
What I'm not sure about is what effect that has on training, I'm currently running a long training with 4 GPUs and pipeline_stages=2 and it seems to me that it causes concept bleeding but I can't know for sure.
The speed increase is not, however x, in my case the pipeline stages=2 gives a more or less 1.8 speed increase. So I wouldn't recommend using multi GPUs to speed up things unless you are in a hurry.

**Comment by tdrussell** [comment]
The global batch size is the per GPU batch size, multiplied by the data parallel degree.

If you have 4 GPUs, with pipeline_stages=2, you have DP degree 2 and PP degree 2. In other words, two instances of the model, but each instance is divided layer-wise across two GPUs.

Increasing data parallel degree won't decrease the per-step time. It will just cause larger global batches because now multiple instances of the model are processing in parallel.

Also keep in mind that with pipeline parallelism, you need to raise gradient_accumulation_steps so that the GPUs will be active at the same time (that's what the pipelining in PP is about). If you are using PP with 1 gradient accumulation step, then only one GPU is active at a time, as they handoff the single micro-batch between them (no pipelining of multiple micro-batches).

**Comment by Sofocletus** [comment]
Thank you for the answer, although I'm still not grasping the whole picture.
Could you recommend a link where I can study the matter? Because to be honest "In other words, two instances of the model, but each instance is divided layer-wise across two GPUs" is going over my head at the moment. I'd like to know more to understand what effect that has on training.

In any case I think the need to change the Gradient Accumulation value should be mentioned in the Parallelism section of the Readme because it seems like an important detail with a large impact on training time (I don't know if it affects quality too).
So for example, with my last training using 4 GPUs and pipeline_stages=2 the Gradient Accumulation should be 2 or 4?
